In [1]:
# Importing libraries
import numpy as np
import pandas as pd
from mlxtend.frequent_patterns import apriori as ap
from mlxtend.frequent_patterns import association_rules as ar

In [2]:
# Reading the file
data = pd.read_csv(r'C:\Users\WIN8\Desktop\PGDS 2020\Datasets\Apriori\Online Retail.csv', encoding = 'unicode_escape')

In [3]:
len(data)

541909

In [4]:
data.head(10)

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
5    536365     22752         SET 7 BABUSHKA NESTING BOXES         2   
6    536365     21730    GLASS STAR FROSTED T-LIGHT HOLDER         6   
7    536366     22633               HAND WARMER UNION JACK         6   
8    536366     22632            HAND WARMER RED POLKA DOT         6   
9    536367     84879        ASSORTED COLOUR BIRD ORNAMENT        32   

        InvoiceDate  UnitPrice  CustomerID         Country  
0  01-12-2010 08:26       2.55     17850.0  United Kingdom  
1  01-12-2010 08:26       3.39     17850.0  United Kingdom  
2  01-12-2010 08:26       2.75     17850.0  United Kingdom  
3  01-12-2010 08:26       3.39     17850.0  United Kingdom  
4  01-12-2010 08:26       3.39     17850.0  United Kingdom  
5  01-12-2010 08:26       7.65     17850.0  United Kingdom  
6  01-12-2010 08:26       4.25     17850.0  United Kingdom  
7  01-12-2010 08:28       1.85     17850.0  United Kingdom  
8  01-12-2010 08:28       1.85     17850.0  United Kingdom  
9  01-12-2010 08:34       1.69     13047.0  United Kingdom

The Apriori Algorithm accepts Binary data. Hence we will have to pre-process the data frame, before running the algorithm. We will need Invoice Number and Item Description for the algorithm. Let us first explore the data. 

In [5]:
data.describe()

Quantity      UnitPrice     CustomerID
count  541909.000000  541909.000000  406829.000000
mean        9.552250       4.611114   15287.690570
std       218.081158      96.759853    1713.600303
min    -80995.000000  -11062.060000   12346.000000
25%         1.000000       1.250000   13953.000000
50%         3.000000       2.080000   15152.000000
75%        10.000000       4.130000   16791.000000
max     80995.000000   38970.000000   18287.000000

We observe that numerical variables quantity and unit price have negative values. We can assume that these are sales returns for items bought as these Invoice numbers start with a 'C' which should ideally denote Credit Note. We cannot use these observations for market basket analysis.

In [6]:
# Finding the observations with negative quantity and unit prices
returns = data[data.Quantity < 0]

# Finding the observations with positive quantity and unit prices
sales = data[data.Quantity > 0]

We have subsetted the sales invoices for our analysis. We will henceforth use the sales dataframe for our analysis. We need to remove the spaces if any before or after the Item descriptions.

In [7]:
# Removing spaces
sales.Description = sales.Description.str.lstrip()
sales.Description = sales.Description.str.rstrip()
# Converting Invoice No. to string
sales.InvoiceNo = sales.InvoiceNo.astype('str')
sales.head()

C:\ProgramData\Anaconda3\lib\site-packages\pandas\core\generic.py:5208: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self[name] = value


InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

        InvoiceDate  UnitPrice  CustomerID         Country  
0  01-12-2010 08:26       2.55     17850.0  United Kingdom  
1  01-12-2010 08:26       3.39     17850.0  United Kingdom  
2  01-12-2010 08:26       2.75     17850.0  United Kingdom  
3  01-12-2010 08:26       3.39     17850.0  United Kingdom  
4  01-12-2010 08:26       3.39     17850.0  United Kingdom

In [8]:
# Looking for missing values in the dataset
sales.isnull().sum()

InvoiceNo           0
StockCode           0
Description       592
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     133361
Country             0
dtype: int64

In [9]:
total_sales_quantity = sum(sales.Quantity)
total_sales_amount = np.dot(sales.Quantity,sales.UnitPrice)
total_invoices = sales.InvoiceNo.unique()
unique_items = sales.StockCode.unique()


print('The Total Item Count is: ', total_sales_quantity)
print('---------------------------------------------')
print('The Number of Unique Items are: ', len(unique_items))
print('---------------------------------------------')
print('The Total Sales Amount is: ', round(total_sales_amount,2))
print('---------------------------------------------')
print('The Total Number of Sales Invoices are: ', len(total_invoices))

The Total Item Count is:  5660981
---------------------------------------------
The Number of Unique Items are:  3941
---------------------------------------------
The Total Sales Amount is:  10644560.42
---------------------------------------------
The Total Number of Sales Invoices are:  20728


We do not have the currency details for the sales amount. As we can see that we have a total of 3941 unique items in our data. Now we will create a dataframe with the unique items as the columns and binary values for the items present in each invoice. Also we have observations from different countries. It is meaningful to get association rules for specific countries. We will use United Kingdom here.

In [10]:
# Preparing data for the algorithm
df = (sales[sales['Country'] == 'Germany']
.groupby(['InvoiceNo', 'Description'])['Quantity'].sum().unstack().reset_index().fillna(0).set_index('InvoiceNo'))

In [11]:
# Viewing the data
df.head()

Description  10 COLOUR SPACEBOY PEN  12 COLOURED PARTY BALLOONS  \
InvoiceNo                                                         
536527                          0.0                         0.0   
536840                          0.0                         0.0   
536861                          0.0                         0.0   
536967                          0.0                         0.0   
536983                          0.0                         0.0   

Description  12 IVORY ROSE PEG PLACE SETTINGS  \
InvoiceNo                                       
536527                                    0.0   
536840                                    0.0   
536861                                    0.0   
536967                                    0.0   
536983                                    0.0   

Description  12 MESSAGE CARDS WITH ENVELOPES  12 PENCIL SMALL TUBE WOODLAND  \
InvoiceNo                                                                     
536527                                   0.0                            0.0   
536840                                   0.0                            0.0   
536861                                   0.0                            0.0   
536967                                   0.0                            0.0   
536983                                   0.0                            0.0   

Description  12 PENCILS SMALL TUBE RED RETROSPOT  12 PENCILS SMALL TUBE SKULL  \
InvoiceNo                                                                       
536527                                       0.0                          0.0   
536840                                       0.0                          0.0   
536861                                       0.0                          0.0   
536967                                       0.0                          0.0   
536983                                       0.0                          0.0   

Description  12 PENCILS TALL TUBE POSY  12 PENCILS TALL TUBE RED RETROSPOT  \
InvoiceNo                                                                    
536527                             0.0                                 0.0   
536840                             0.0                                 0.0   
536861                             0.0                                 0.0   
536967                             0.0                                 0.0   
536983                             0.0                                 0.0   

Description  12 PENCILS TALL TUBE SKULLS  ...  YULETIDE IMAGES GIFT WRAP SET  \
InvoiceNo                                 ...                                  
536527                               0.0  ...                            0.0   
536840                               0.0  ...                            0.0   
536861                               0.0  ...                            0.0   
536967                               0.0  ...                            0.0   
536983                               0.0  ...                            0.0   

Description  ZINC  HEART T-LIGHT HOLDER  ZINC  STAR T-LIGHT HOLDER  \
InvoiceNo                                                            
536527                              0.0                        0.0   
536840                              0.0                        0.0   
536861                              0.0                        0.0   
536967                              0.0                        0.0   
536983                              0.0                        0.0   

Description  ZINC BOX SIGN HOME  ZINC FOLKART SLEIGH BELLS  \
InvoiceNo                                                    
536527                      0.0                        0.0   
536840                      0.0                        0.0   
536861                      0.0                        0.0   
536967                      0.0                        0.0   
536983                      0.0                        0.0   

Description  ZINC HEART LATTIC

In [12]:
# Converting all positive values to 1 and everything else to 0
def encode(x):
    if x <= 0:
        return 0
    if x >= 1:
        return 1

df = df.applymap(encode)
df.drop('POSTAGE', inplace=True, axis=1)

An item in the data 'Postage' seems insignificant. Hence we remove it from our analysis.

### Training the Model

In [13]:
# Generating frequent itemsets
frequent_itemsets = ap(df, min_support = 0.05, use_colnames = True)

In [14]:
# Generating Rules
rules = ar(frequent_itemsets, metric = "lift", min_threshold = 2)

In [15]:
# Viewing top 100 rules
rules.head(100)

antecedents  \
0        (PLASTERS IN TIN CIRCUS PARADE)   
1     (PLASTERS IN TIN WOODLAND ANIMALS)   
2        (PLASTERS IN TIN CIRCUS PARADE)   
3    (ROUND SNACK BOXES SET OF 4 FRUITS)   
4        (PLASTERS IN TIN CIRCUS PARADE)   
5   (ROUND SNACK BOXES SET OF4 WOODLAND)   
6     (PLASTERS IN TIN WOODLAND ANIMALS)   
7             (PLASTERS IN TIN SPACEBOY)   
8     (PLASTERS IN TIN WOODLAND ANIMALS)   
9   (ROUND SNACK BOXES SET OF4 WOODLAND)   
10              (WOODLAND CHARLOTTE BAG)   
11         (RED RETROSPOT CHARLOTTE BAG)   
12   (ROUND SNACK BOXES SET OF 4 FRUITS)   
13  (ROUND SNACK BOXES SET OF4 WOODLAND)   
14                  (SPACEBOY LUNCH BOX)   
15  (ROUND SNACK BOXES SET OF4 WOODLAND)   
16              (WOODLAND CHARLOTTE BAG)   
17  (ROUND SNACK BOXES SET OF4 WOODLAND)   

                             consequents  antecedent support  \
0     (PLASTERS IN TIN WOODLAND ANIMALS)            0.115974   
1        (PLASTERS IN TIN CIRCUS PARADE)            0.137856   
2    (ROUND SNACK BOXES SET OF 4 FRUITS)            0.115974   
3        (PLASTERS IN TIN CIRCUS PARADE)            0.157549   
4   (ROUND SNACK BOXES SET OF4 WOODLAND)            0.115974   
5        (PLASTERS IN TIN CIRCUS PARADE)            0.245077   
6             (PLASTERS IN TIN SPACEBOY)            0.137856   
7     (PLASTERS IN TIN WOODLAND ANIMALS)            0.107221   
8   (ROUND SNACK BOXES SET OF4 WOODLAND)            0.137856   
9     (PLASTERS IN TIN WOODLAND ANIMALS)            0.245077   
10         (RED RETROSPOT CHARLOTTE BAG)            0.126915   
11              (WOODLAND CHARLOTTE BAG)            0.070022   
12  (ROUND SNACK BOXES SET OF4 WOODLAND)            0.157549   
13   (ROUND SNACK BOXES SET OF 4 FRUITS)            0.245077   
14  (ROUND SNACK BOXES SET OF4 WOODLAND)            0.102845   
15                  (SPACEBOY LUNCH BOX)            0.245077   
16  (ROUND SNACK BOXES SET OF4 WOODLAND)            0.126915   
17              (WOODLAND CHARLOTTE BAG)            0.245077   

    consequent support   support  confidence      lift  leverage  conviction  
0             0.137856  0.067834    0.584906  4.242887  0.051846    2.076984  
1             0.115974  0.067834    0.492063  4.242887  0.051846    1.740427  
2             0.157549  0.050328    0.433962  2.754455  0.032057    1.488330  
3             0.115974  0.050328    0.319444  2.754455  0.032057    1.298977  
4             0.245077  0.056893    0.490566  2.001685  0.028470    1.481887  
5             0.115974  0.056893    0.232143  2.001685  0.028470    1.151290  
6             0.107221  0.061269    0.444444  4.145125  0.046488    1.607002  
7             0.137856  0.061269    0.571429  4.145125  0.046488    2.011670  
8             0.245077  0.074398    0.539683  2.202098  0.040613    1.640006  
9             0.137856  0.074398    0.303571  2.202098  0.040613    1.237951  
10            0.070022  0.059081    0.465517  6.648168  0.050194    1.739959  
11            0.126915  0.059081    0.843750  6.648168  0.050194    5.587746  
12            0.245077  0.131291    0.833333  3.400298  0.092679    4.529540  
13            0.157549  0.131291    0.535714  3.400298  0.092679    1.814509  
14            0.245077  0.070022    0.680851  2.778116  0.044817    2.365427  
15            0.102845  0.070022    0.285714  2.778116  0.044817    1.256018  
16            0.245077  0.063457    0.500000  2.040179  0.032354    1.509847  
17            0.126915  0.063457    0.258929  2.040179  0.032354    1.178139

### Making Recommendations

In [16]:
rules[ (rules['lift'] >= 3) &
      (rules['confidence'] >= 0.5) ]

antecedents  \
0        (PLASTERS IN TIN CIRCUS PARADE)   
7             (PLASTERS IN TIN SPACEBOY)   
11         (RED RETROSPOT CHARLOTTE BAG)   
12   (ROUND SNACK BOXES SET OF 4 FRUITS)   
13  (ROUND SNACK BOXES SET OF4 WOODLAND)   

                             consequents  antecedent support  \
0     (PLASTERS IN TIN WOODLAND ANIMALS)            0.115974   
7     (PLASTERS IN TIN WOODLAND ANIMALS)            0.107221   
11              (WOODLAND CHARLOTTE BAG)            0.070022   
12  (ROUND SNACK BOXES SET OF4 WOODLAND)            0.157549   
13   (ROUND SNACK BOXES SET OF 4 FRUITS)            0.245077   

    consequent support   support  confidence      lift  leverage  conviction  
0             0.137856  0.067834    0.584906  4.242887  0.051846    2.076984  
7             0.137856  0.061269    0.571429  4.145125  0.046488    2.011670  
11            0.126915  0.059081    0.843750  6.648168  0.050194    5.587746  
12            0.245077  0.131291    0.833333  3.400298  0.092679    4.529540  
13            0.157549  0.131291    0.535714  3.400298  0.092679    1.814509

### Interpretation of the results

Antecedent Support for rule[0] -> 11.6% of the total transactions under our analysis contain the antecedent item 'PLASTERS IN TIN CIRCUS PARADE'.

Consequent Support for rule[0] -> 13.79% of the total transactions under our analysis contain the consequent item 'PLASTERS IN TIN WOODLAND ANIMALS'.

Support of the rule[0] -> 6.78% of the total number of transactions under our analysis contains these 2 items.

Confidence of rule[0] -> Out of all the transactions that contain the antecedent ('PLASTERS IN TIN CIRCUS PARADE'), 58.49% of them contains the consequent (''PLASTERS IN TIN WOODLAND ANIMALS').

Lift of rule[0] -> The probability of purchasing the antecedent and the consequent together is 4.24 times more than what it would be if they were independent of one another.